src/Illuminate/Routing/AbstractRouteCollection.php

```diff
@@ -0,0 +1,215 @@
+<?php
+
+namespace Illuminate\Routing;
+
+use ArrayIterator;
+use Countable;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use Illuminate\Support\Str;
+use IteratorAggregate;
+use Symfony\Component\HttpKernel\Exception\MethodNotAllowedHttpException;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Component\Routing\Matcher\Dumper\CompiledUrlMatcherDumper;
+use Symfony\Component\Routing\RouteCollection as SymfonyRouteCollection;
+
+abstract class AbstractRouteCollection implements Countable, IteratorAggregate, RouteCollectionInterface
+{
+    /**
+     * Handle the matched route.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Illuminate\Routing\Route|null  $route
+     * @return \Illuminate\Routing\Route
+     *
+     * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException
+     */
+    protected function handleMatchedRoute(Request $request, $route)
+    {
+        if (! is_null($route)) {
+            return $route->bind($request);
+        }
+
+        // If no route was found we will now check if a matching route is specified by
+        // another HTTP verb. If it is we will need to throw a MethodNotAllowed and
+        // inform the user agent of which HTTP verb it should use for this route.
+        $others = $this->checkForAlternateVerbs($request);
+
+        if (count($others) > 0) {
+            return $this->getRouteForMethods($request, $others);
+        }
+
+        throw new NotFoundHttpException;
+    }
+
+    /**
+     * Determine if any routes match on another HTTP verb.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return array
+     */
+    protected function checkForAlternateVerbs($request)
+    {
+        $methods = array_diff(Router::$verbs, [$request->getMethod()]);
+
+        // Here we will spin through all verbs except for the current request verb and
+        // check to see if any routes respond to them. If they do, we will return a
+        // proper error response with the correct headers on the response string.
+        $others = [];
+
+        foreach ($methods as $method) {
+            if (! is_null($this->matchAgainstRoutes($this->get($method), $request, false))) {
+                $others[] = $method;
+            }
+        }
+
+        return $others;
+    }
+
+    /**
+     * Determine if a route in the array matches the request.
+     *
+     * @param  array  $routes
+     * @param  \Illuminate\Http\Request  $request
+     * @param  bool  $includingMethod
+     * @return \Illuminate\Routing\Route|null
+     */
+    protected function matchAgainstRoutes(array $routes, $request, $includingMethod = true)
+    {
+        [$fallbacks, $routes] = collect($routes)->partition(function ($route) {
+            return $route->isFallback;
+        });
+
+        return $routes->merge($fallbacks)->first(function (Route $route) use ($request, $includingMethod) {
```

Comment:
do we want the typechck here, or should we remove it like the rest of the code?

Reply:
This part hasn't been touched so I'm gonna leave it be.